walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,191 @@
+   *     Includes name, duration, and collection of attendees
+   * @param ignoreOptional false if optional attendees should be accommodated for scheduling. True
+   *     otherwise
+   * @return A collection of time ranges when the new meeting can occur without conflict
+   */
+  public List<TimeRange> queryToggleOptional(
+      Collection<Event> events, MeetingRequest request, boolean ignoreOptional) {
+
+    // Get the current attendees, including/excluding the optional ones if specified
+    // Use HashSet to prevent duplicates
+    HashSet<String> attendees = new HashSet<>(request.getAttendees());
+    if (!ignoreOptional) {
+      attendees.addAll(request.getOptionalAttendees());
+    }
+
+    // Filter the data to remove events with no shared attendees with the meeting request
+    // Sort the events from earliest start time to latest
+    List<TimeRange> eventTimes = eventsToSortedTimes(events, attendees);
+
+    // Merge the event times into a sorted list of non-overlapping events
+    List<TimeRange> availableTimes = generateSuggestedTimes(eventTimes, request.getDuration());
+
+    return availableTimes;
+  }
+
+  /**
+   * Converts a list of events to a sorted list of time ranges for the events. Will remove events
+   * that are irrelevant (i.e. the event attendees don't match the requestedAttendees, or the event
+   * only contains optional attendees and ignoreOptional is true). Will sort by start time, from
+   * earliest in the day to latest
+   *
+   * @param events list of Event objects that the meeting request must work around
+   * @param requestedAttendees list of attendees (optional and required) requested for the new
+   *     meeting
+   * @return a list of time ranges of events that must be worked around
+   */
+  private List<TimeRange> eventsToSortedTimes(
+      Collection<Event> events, Collection<String> requestedAttendees) {
+    return events.stream()
+        .filter(
+            e -> {
+              HashSet<String> sharedAttendees = new HashSet<>(e.getAttendees());
+              sharedAttendees.retainAll(requestedAttendees);
+              return sharedAttendees.size() != 0;
+            })
+        .map(Event::getWhen)
+        .sorted(TimeRange.ORDER_BY_START)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Takes in a list of all events and outputs a list of available time ranges that can accommodate
+   * the requested meeting duration
+   *
+   * @param eventTimes a sorted list of event time ranges (from earliest start time to latest start
+   *     time)
+   * @param requestedDurationMinutes a length of time given in minutes for the requested meeting.
+   *     Must be > 0.
+   * @return a sorted list of non-overlapping time ranges (from earliest start time to latest start
+   *     time)
+   */
+  private List<TimeRange> generateSuggestedTimes(List<TimeRange> eventTimes, long requestedDurationMinutes) {
+    // Create empty list to store busy periods
```

Comment:
1. Nit: Did you mean available times?
2. FFT: Having comments describing what the code is doing often leads to that kind of "error" because you end up violating the DRY principle. The name of your variable, `availableTimes`, already describes what it is meant for (kudos for picking that name), and `= new ArrayList<>()` tells the reader you are creating an empty list...

Reply:
I agree completely - I think leaving more meaningful, sparse comments is probably a good focus for me. 

Definitely going to review the training we had on meaningful comments - often I leave these just to keep myself organized, but I should take the time to remove them so the code becomes more maintainable